### ISSUE TITLE
LC elastic model

### ISSUE THREAD

**Issue by sebhinck** [pism/pism#424]
When enabling the LC elastic model I found strange artifacts in the uplift rate (see Figure) that seem to originate from some kind of unsymmetrical problem.
![uplift_lc_elastic](https://user-images.githubusercontent.com/25295341/48630786-5f31f380-e9bd-11e8-8aba-102986942ccc.png)

This is why I had a closer look on how the elastic deformation is calculated. According to Bueler et al 2007 it is calculated as the convolution of the load with the Green's function, which is calculated once at the initialization of the LingleClarkSerial model (`m_load_response_matrix` line 218ff `src/earth/LingleClarkSerial.cc`). This matrix (see figure below) is only the upper right quadrant of the radially symmetrical Green's function (with the point load located in the origin). This matrix contains all information needed, but then the convolution function needs to take care of this.

![load_response_matrix](https://user-images.githubusercontent.com/25295341/48630826-7b359500-e9bd-11e8-9a34-ee0f376effe7.png)

`conv2_same` in `src/earth/matlablike.cc` line 26ff does only integrate the product of Green's function and load over the south-western (lower left) quadrant and thus neglects contributions of masses in the other quadrants. One simple but very computational demanding solution would be:
```
void conv2_same(Vec vA, int mA, int nA,  Vec vB, int mB, int nB,
                Vec vresult) {

  pism::petsc::VecArray2D
    A(vA, mA, nA),
    B(vB, mB, nB),
    result(vresult, mA, nA);

  for (int j=0; j < nA; j++) {
    for (int i=0; i < mA; i++) {
      double sum = 0.0;

      for (int r = 0; r < mA; r++) {
        for (int s = 0; s < nA; s++) {
          sum += A(r, s) * B(std::abs(i - r), std::abs(j - s));
        }
      }
      result(i, j) = sum;
    }
  }
}
```
Compare the elastic deformation fields with and without this modification. Note the deformation being zero south and west of the ice sheet, which it shouldn't.

![elastic_compare](https://user-images.githubusercontent.com/25295341/48630853-87b9ed80-e9bd-11e8-9c01-4bc39c6b87fe.png)

I think there exist more elaborate functions to calculate the convolution based on FFT which are more efficient than this one here, and which is also mentioned in Bueler et al 2007.

Please correct me if I misunderstood something and I am wrong.

I also have a question regarding #183 (make the on/off state of the elastic part of the "-bed_def lc" earth model configurable #183). What is the reason for disabling this part of the model by default though it is an Elastic Lithosphere Relaxing Asthenosphere (ELRA) model. This should probably be mentioned more explicitly in the user manual.

**Comment by sebhinck** [comment]
I have done the FFTW implementation here: https://github.com/sebhinck/pism-pub/commits/lc_elastic_fftw

This results in the same (O(1e-12m)) elastic deformation as the code snippet I proposed above, but is much faster. I hope I haven't overseen any required scaling of the result. If not the elastic deformation has a much higher contribution to the total bed deformation as before (~15%).

To work properly `m_load_response_matrix` needed to be adapted (see Figure).
![lc_response_matrix_fftw](https://user-images.githubusercontent.com/25295341/48662528-7e935400-ea83-11e8-8b29-c09f307396cd.png)

To output the load response matrix and the elastic deformation I wrote short code snippets:
[response_matrix_out_patch.txt](https://github.com/pism/pism/files/2591937/response_matrix_out_patch.txt)
[elastic_bed_def_out_patch.txt](https://github.com/pism/pism/files/2591940/elastic_bed_def_out_patch.txt)

**Comment by ckhroulev** [comment]
@sebhinck Thanks for reporting this! (And for the fixes, too!)

This will likely be the first or second bug fix of the new PISM release (`v1.1`).

**Comment by sebhinck** [comment]
Ok, thank you @ckhroulev .
Although probably not a lot people using this feature it should also be backported to older versions I think. The impact on the bed deformation can be quite immense (in my example above ~200m instead of 60m). Don't you think?

Since it is not really computationally expensive but can have a big impact on the ice dynamics it should probably also be turned on by default.

**Comment by aaschwanden** [comment]
@sebhinck There are many setups where bed deformation is undesirable (i.e. synthetic geometry experiments). I think the default should remain to be off.

**Comment by talbrecht** [comment]
I though @sebhinck was talking about the elastic part of the ELRA, if `-bed_def lc` is activated, or am I wrong? I have to admit that I wasn't aware of the elastic part switched off for quite a while...

**Comment by ckhroulev** [comment]
@aaschwanden Indeed, I believe the question is not whether bed deformation should be "on" by default, but whether the elastic part of the Lingle-Clark model should be "on" by default when `-bed_def lc` is set.

@talbrecht Note that the model implemented in PISM is **not** the ELRA (it is very similar, but different). See *E. Bueler, C. S. Lingle, and J. A. Kallen-Brown. Fast computation of a viscoelastic deformable Earth model for ice sheet simulation. Ann. Glaciol., 46:97–105, 2007* for details (the Discussion, specifically). The elastic part of the model has been "off" by default since 4694d2f504, as far as I can tell, i.e. since July 2007 at the least.

@sebhinck I'm OK with back-porting this to v1.0 (and maybe even v0.7, but you'll have to convince me that it's worth the time and effort). Once the elastic part of the LC model is fixed and tested we could enable it by default in future PISM versions. Changing the default setting in older PISM versions, however, is a bad idea.

**Comment by talbrecht** [comment]
I agree, but from a user point of view the corresponding description of the LC model in the manual (Earth deformation models) suggest "fast computation of a viscoelastic deformable Earth model". If the elastic part is switched off by default I would recommend to emphasize this in the manual. I also agree that backporting to previous stable versions is not a good idea.

**Comment by sebhinck** [comment]
Yes, I was talking about enabling the elastic part of the LC model (only when using this beddef model) @aaschwanden @talbrecht.



@ckhroulev As mentioned in the paper the elastic component gives about 12-14% contribution to total earth deformation, which might be critically, especially when running transient simulations with rapid deglaciations. 

Most PISM users I know about still work with PISM v0.7 (for compareability or also stability reasons), but I don't know how this is for the entire PISM community. If this fix is back-ported depends on in how far older PISM versions are being maintained. I see this as a serious bug that affects all users who enable this part of the model. 

I wouldn't change the default settings in the old PISM versions (especially not if the fix is not back-ported), and the same for the current version if there are good reasons for not enabling it. But in this case it should be mentioned in the user manual, because most people I talked about this issue were quite surprised that it was disabled.

**Comment by aaschwanden** [comment]
OK, I misunderstood. Yes we could enable this in the next stable version. I hope to release the next version in less than a year, hopefully in spring or early summer. We have a bunch of new physics that is currently  being developed but didn't make the cut yet.

**Comment by flapo099** [comment]
Thank you very much for this issue being reported. We are using heavily the -bed_def lc as we are looking into the self-graviational effect of ice and land on sea level. This fix for the v1.1 will be enough for us as we are using the new 2D sea level fonction.

**Comment by 45jan** [comment]
Thanks for solving the issue. Am I right, that it has not been fixed for v1.1.3? I consider copying @sebhinck's suggestions, but since I am new to pism I am afraid to introduce other bugs which I might not even notice.
Are there reasons (other than time/resources) for that the changes have not been merged so far?

**Comment by sebhinck** [comment]
@45jan By just cherry-picking (https://git-scm.com/docs/git-cherry-pick) my three commits (cb7390e42ad48e818862b3c5bd006a01efb072ec, b80e6058fa891a1908a8ebc69decf145e636ce8c and 48625abbb51b560f4359796a21a85feb58806ee5) you should in general not introduce any other bugs (given there are no bugs in my implementation...).

**Comment by 45jan** [comment]
Thank you, @sebhinck. I'm quite new to git, too, and didn't know about this option. I will do so, I guess I'll be fine.

**Comment by sebhinck** [comment]
@tkleiner just made me aware of a small bug in my code. Thank you!

Instead of writing the elastic deformation into the local variable of the function that is meant for output `dE`, I wrote it to the global variable `m_Ue`. In the end this won't change the behavior at all because `dE` is basically just a reference to the global `m_Ue` in this context. 
This commit just cleans the code and gets rid of the compiler warning that dE is not used.

See my new commit: https://github.com/pism/pism/commit/94f47ce603cb2e6d23dd7ac2c0caaf151b8be59e

**Comment by tkleiner** [comment]
@sebhinck Thank you! 

The following worked for me within my v1.1.3 working copy:
```
git remote add sebhinck https://github.com/sebhinck/pism-pub.git
git fetch sebhinck
git cherry-pick -x cb7390e 
git cherry-pick -x b80e605
git cherry-pick -x 48625ab
git cherry-pick -x 94f47ce
```
Note, 'sebhinck' is an arbitrary name for the remote. You could chose something else.

**Comment by ckhroulev** [comment]
This issue is fixed in the new PISM release (v1.2).
